ls1intum/Artemis
# `General`: Add dark mode ☾ ✨

### Checklist
#### General


#### Client

### Motivation and Context



Dark Modes are commonly available in most software used by programmers today.
Operating Systems, IDEs and browsers all come with some sort UI mode with negative polarity.

As Artemis is mostly used for programming exercises in CS courses and thus is used by a lot of programmers both on the student and the instructor side, we should provide a native dark mode as well.

Closes #4404.

### Description


#### Application (TS / HTML)

Added a new theme service that injects a theme override stylesheet to the page ``<head>`` if the dark mode shoud be applied. The theme service has been designed to support more than one additional theme in the future.
The service automatically infers the preferred UI mode from OS / browser settings on first launch, but stores the user preference in local storage from then on.

Added a theme switch component to the navbar to allow for easy and convenient theme switching.
It also contains a popover with a few hints and a big flag that marks this feature as experimental.

#### Styles

The dark mode was implemented by modifying the variables of bootstrap before importing it.
However (or rather, sadly), we have lots and lots of local color definitions in component stylesheets that do not re-use bootstrap colors.

Therefore, a lot of these colors in component stylesheets have been replaced with CSS variables that get different values for both themes. For each theme, there is now a "variables" SCSS file where these colors are defined.
In this first version, there are a lot of variables in there. Future work has to be done to reduce the amount of variables, re-use colors where possible, and reduce the amount of used colors throughout the app in general (we really don't need 38293842 shades of gray everywhere).

The default theme is (like it was before) injected by Angular by default. The dark theme is injected as an overriding stylesheet when requested.

Apollon has been themed as well. A huge thank you to @martindevtum who made easy theming possible in Apollon.

### Steps for Testing



- On first launch of PR: Make sure that the app automatically shows the popover and syncs the theme to your OS setting by default. If you need to test it again, clear local storage for the webapp.
    - Please test this with both themes; at least in Firefox, you can select which preferred theme the browser should report to the page in dev tools:
     ![grafik](https://user-images.githubusercontent.com/23171488/165314960-474cc5c9-949b-4c6b-8240-1ecb6a12a763.png)
    - If you do it with... (Please test both)
        - ...the light theme, it should automatically show you a popover offering you a "Enable now" button to apply the dark mode, alongside a "No thanks" link that will store the light preference in local storage. 
        - ...the dark theme, it should inform you that it auto-enabled the dark mode based on your preferences and that you can disable it at any time. Please check that the "Got it!" link closes the popover.
- Make sure that you can switch between themes by clicking the icon in the navbar at all times. Verify that this disables the OS-sync feature automatically
- Make sure that your selection survives reload and browser restarts
- Make sure that you can enable and disable the OS-sync feature at any time. If it's enabled, it should sync with the system. In Firefox, you can test this as shown above. If it's disabled, it should keep the currently active theme.
- Probably the biggest thing: In general, this is an experimental feature and we need to make sure that it works everywhere. I expect this to be a longer process where we collect issues over time. Therefore, it's not expected that you test each and every page right now. We should do so in a testing session. However, if there are elements that you know and that are important etc, please go ahead and check if they look okay in both themes.

### Code review

This PR is ultra large. However, a lot of the changes are related to fixing styles and replacing colors with variables.

The most important files to review are the following as they contain the most prominent business logic changes and/or new components:
```
src/main/webapp/app/core/theme/*
src/main/java/de/tum/in/www1/artemis/service/PlantUmlService.java
src/main/java/de/tum/in/www1/artemis/web/rest/PlantUmlResource.java
src/main/resources/config/liquibase/changelog/20220503131200_changelog.xml
```

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| theme-switch.component.ts | 58% |  89% |
| theme.service.ts | 75% | 91% |

### Screenshots



https://user-images.githubusercontent.com/23171488/165316241-244bf6bf-b5e0-46c0-9a12-24c17020196b.mov
